Source/DafnyCore/AST/IndentationFormatter.cs

```diff
@@ -6,6 +6,45 @@
 
 namespace Microsoft.Dafny;
 
+/** The first Dafny formatter
+ * 
+ * We want the guarantee that the reprinted program should be the same as the previous one, modulo whitespace.
+ * Our approach in this formatter consists of three phases:
+ * 1. Convert tokens to a double linked list so that we can traverse them from the beginning to the end and store the first token in the program (Scanner.frame, and all AST nodes)
+ * 2. Traverse the program on pre-resolved nodes, and decide on the indentation of each token when it's associated to a declaration, a node, a statement, etc.
```

Comment:
> Seems like there's a single pass over the AST that modifies the posToIndentBefore, posToIndentLineBefore and posToIndentAfter maps, and then a loop through all tokens which uses these maps to reindent them.

Yes, good summary.

> What would be the difference with something that looks more like a printer, where the iteration over the AST immediately produces the output string?

I do consider to make a token-based printer in the future so that we can deal with automated newlines.

The biggest difference with printers is that you will not be able to guarantee that tokens are printed the same. So you risk producing incorrect output if your code is insufficiently tested. You might forget trivia, comments, existing alignments, etc. A printer might degroup cases that were degroupped during parsing. Moreover, if you add a new construct to the language, you'll have to add a new printer expression for that, otherwise you will simply break the formatter. Token-based formatting does not suffer this. It offers a default formatting even for expressions that it does not know yet, and guarantees that your original file is still there, modulo some whitespace reformatting.

Reply:
Okay. I think having this two phase process of having the AST produce formatting hints and then printing all existing tokens using those seems good then, given that it gives this 'no text is lost' guarantee and the ability to work with unknown pieces of grammar.

I do not yet understand the structure of the formatting hints though. Suppose tokens were just lexical tokens, then I would think the formatting hints would be additions and removals of whitespace/newline tokens at specific points in the list of all tokens.

Also, why are trivia-tokens attached to non-trivia tokens?

Reply:
> I do not yet understand the structure of the formatting hints though. Suppose tokens were just lexical tokens, then I would think the formatting hints would be additions and removals of whitespace/newline tokens at specific indices in the list of all tokens.

That would be true if comments were themselves tokens, and trivia was themselves tokens, but we haven't implemented it that way.

> Also, why are trivia-tokens attached to non-trivia tokens?

This was a decision taken in a previous PR that attached the trivia to every token. (trivia themselves are not tokens). That way, tokens represent something concrete, whereas trivia represents something that, besides formatting, is not use elsewhere.

Reply:
Would it be possible to replace 

```c#
  private readonly Dictionary<int, int> posToIndentBefore = new();
  private readonly Dictionary<int, int> posToIndentLineBefore = new();
  private readonly Dictionary<int, int> posToIndentAfter = new();
```

with

```c#
private readonly Dictionary<IToken, (string leadingTrivia, string trailingTrivia)> 
  triviaReplacements = new();
```

To me that seems to be the minimal data needed to allow you to ignore tokens for which no formatting is defined, and guarantee that all semantic tokens end up in the formatted output.